proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,478 @@
+platform.
+
+### Replacing environment variables in Package manifests
+
+A lot of packages are using environment variables in their `Package.swift` to
+configure their package. This has various reasons such as optional dependencies
+or setting certain defines for local development. Using environment variables
+inside `Package.swift` is not officially supported and with stricter sandboxing
+rules might break. Package traits provide provide a blessed way to configure a
+package.
+
+## Proposed solution
+
+This proposal introduces a new configuration for packages called _package
+traits_. Package authors can define a set of traits in their `Package.swift`
+that their package offers which provide a way to express conditional compilation
+and optional dependencies. Furthermore, a set of default enabled traits can be
+specified.
+
+```swift
+let package = Package(
+    name: "Example",
+    traits: [
+        "Foo",
+        Trait(
+            name: "Bar",
+            enabledTraits: [ // Other traits that are enabled when this trait is being enabled
+                "Foo"
+            ]
+        )
+    ],
+    defaultTraits: [ // The set of default enabled traits
+        "Foo",
+    ],
+    /// ...
+)
+```
+
+When depending on a package a set of enabled traits can be passed and wether the
+default set of traits are disabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: ["SomeTrait"],
+            disableDefaultTraits: true
+        )
+    ),
+]
+```
+
+Another common scenario is to enable a trait of a dependency only when a trait
+of the package is enabled. The below example enables the `SomeOtherTrait` when
+the `Foo` trait of this package is enabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: [
+                "SomeTrait",
+                EnabledTrait("SomeOtherTrait", condition: .when(traits: ["Foo"])),
+            ]
+        )
+    ),
+]
+```
+
+Conditional dependencies are specified per target and extend the current
+`condition` syntax which is used for specifying platform dependent dependencies.
+
+```swift
+targets: [
+    .target(
+        name: "SomeTarget",
+        dependencies: [
+            .product(
+                name: "SomeProduct",
+                package: "SomePackage",
+                condition: .when(traits: ["Foo"])
+            ),
+        ]
+    )
+]
+```
+
+Lastly, code can be conditionally compiled by checking if a trait is enabled.
+This can be used for both optional dependencies by surrounding the `import`
+statements in a trait check and for regular code where you want to modify its
+behaviour depending on the enabled traits.
+
+```swift
+#if TRAIT_Foo
```

Comment:
I agree. I started working on `#if trait(Foo)` support but wanted to draft up the proposal with the current state already.

Reply:
I agree with Franz that this isn't critical for the proposal as it requires additional compiler work, but certainly deserves to be mentioned in "Future Directions".

Reply:
This is exactly what compilation conditions are, so it seems odd to me that we'd want another syntax for them. Is there any particular reason that we need to prefix it at all? Maybe needs to be converted to uppercase + underscored so that this would be `#if FOO` though.

There's also C/C++ to consider, but I also think just passing a macro definition through would be fine there too.

Reply:
I have already called out a potential future direction of `#if trait(Foo)`. For now I would like to keep it as `#if TRAIT_Foo` and pitch it. I assume a lot of people of opinions about this syntax.

Reply:
> Is there any particular reason that we need to prefix it at all?

I guess the only reason is to avoid collisions with non-trait compilation conditions, but you're right — this is what they're for anyway, so maybe we don't need a prefix at all?

Reply:
I question if there's also a danger of running into confusion versus `#if hasFeature(Foo)` (which I still generally end up discarding in favor of the `#if $Foo` syntax, to add yet another layer of legacy confusion).